test_palindrome.py

```diff
@@ -0,0 +1,151 @@
+import pytest
+from collections import namedtuple
+from palindrome import is_palindrome
+
+
+@pytest.mark.parametrize("case_sensitive", [True, False])
+@pytest.mark.parametrize("whitespace_sensitive", [True, False])
+@pytest.mark.parametrize("punctuation_sensitive", [True, False])
+def test_non_string_data_raises_valueerror(
+        case_sensitive: bool,
+        whitespace_sensitive: bool,
+        punctuation_sensitive: bool,
+) -> None:
+    TestCaseNoOptions = namedtuple(
+        typename='TestCaseNoOptions',
+        field_names=["expectation", "data"]
+    )
+    test_cases = [
+        TestCaseNoOptions(expectation="Expected string, got <class 'NoneType'>!", data=None),
+        TestCaseNoOptions(expectation="Expected string, got <class 'bytes'>!", data=b""),
+        TestCaseNoOptions(expectation="Expected string, got <class 'bytearray'>!", data=bytearray()),
+        TestCaseNoOptions(expectation="Expected string, got <class 'tuple'>!", data=tuple()),
+        TestCaseNoOptions(expectation="Expected string, got <class 'list'>!", data=[]),
+        TestCaseNoOptions(expectation="Expected string, got <class 'int'>!", data=0),
+        TestCaseNoOptions(expectation="Expected string, got <class 'float'>!", data=0.0),
+    ]
+    for expectation, data in test_cases:
+        with pytest.raises(ValueError) as value_error:
+            is_palindrome(data, case_sensitive, whitespace_sensitive, punctuation_sensitive)
+        assert str(value_error.value) == expectation, \
+            f"Expected '{expectation}', got '{str(value_error.value)}' for '{data=}' and "\
+            f"options '{case_sensitive=}', '{whitespace_sensitive=}', '{punctuation_sensitive=}'!"
+
+
+@pytest.mark.parametrize("case_sensitive", [True, False])
+@pytest.mark.parametrize("whitespace_sensitive", [True, False])
+@pytest.mark.parametrize("punctuation_sensitive", [True, False])
+def test_empty_string_is_a_palindrome(
+        case_sensitive: bool,
+        whitespace_sensitive: bool,
+        punctuation_sensitive: bool,
+) -> None:
+    expectation, data = True, ""
```

Comment:
W takim wypadku nie tworzyłbym nowych zmiennych, tylko przekazał je na sztywno. Bo w sumie tutaj nie są to nawet zmienne ale stałe. Chociaż widzę, że potem tego używasz w opisie błędu jesli wystąpi `AssertionError`.

Ja bym się skłaniał jednak do użycia stałych. Wydaje mi się, że w tym wypadku nazwa testu jest na tyle precyzyjna, że programista poradzi sobie ze zrozumieniem błędu.